_posts/2016-10-05-writing-fast-rails.md

```diff
@@ -0,0 +1,133 @@
+---
+layout: post
+title:  "Tips for Writing Fast Rails: Part 1"
+date: 2016-10-05 08:37:00
+categories: ["performance", "rails"]
+author: "etagwerker"
+---
+
+[Rails](http://rubyonrails.org/) is a powerful framework. You can write a lot  
+of features in a short period of time. In the process you can easily
+write code that **performs poorly**.
+
+At [Ombu Labs](http://www.ombulabs.com) we like to [maintain Ruby on Rails
+applications](http://www.ombulabs.com/blog/tags/maintenance). In the process
+of maintaining them, adding features and fixing bugs, we like to improve
+the code and its performance (because we are [good *boy scouts*](http://programmer.97things.oreilly.com/wiki/index.php/The_Boy_Scout_Rule)!)
+
+Here are some tips based on our experience.
+
+## Prefer `where` instead of `select`
+
+When you are performing a lot of calculations, you should load as little as
+possible into memory. **Always** prefer a SQL query vs. an object's method call.
+
+<!--more-->
+
+With [ActiveRecord](http://guides.rubyonrails.org/active_record_basics.html), it's
+easy to forget which methods load
+[ActiveRecord::Base](https://github.com/rails/rails/blob/master/activerecord/lib/active_record/base.rb)
+objects into memory and which perform a simple query instead.
+
+**The bigger the table, the slower the object load**. If you have a table with 80
+columns (**sigh!**), loading each record will take a lot longer than a table
+with 3 columns. So, you must  avoid object loads as much as possible. Only load
+objects into memory when you **really need them**.
+
+For example:
+
+    shop_ids.map do |shop_id|
+      products.select { |p| p.shop_id == shop_id }.first
+    end
+
+`select` will load all the `products` into memory and then compare the ids.
+This will be slower than just using `where`.
+
+This will be much faster:
+
+    shop_ids.map do |shop_id|
+      products.where(shop_id: shop_id).first
+    end
+
+Because it will perform the query and **only after the query returns** it will
+load the objects into memory.
+
+## Prefer `pluck` instead of `map`
+
+If you are interested in only a few values per row, you should use `pluck`
+instead of `map`.
+
+For example:
+
+    Order.where(number: 'R545612547').map &:id
+    # Order Load (5.0ms)  SELECT `orders`.* FROM `orders` WHERE `orders`.`number` = 'R545612547' ORDER BY orders.created_at DESC
+    => [1]
+
+As with `select`, `map` will load the `order` into memory and it will get the
+`id` attribute.
+
+Using `pluck` will be faster, because it doesn't need to load an entire object
+into memory.
+
+So this will be much faster:
+
+    Order.where(number: 'R545612547').pluck :id
+    # SQL (0.8ms)  SELECT `orders`.`id` FROM `orders` WHERE `orders`.`number` = 'R545612547' ORDER BY orders.created_at DESC
+    => [1]
+
+For this particular case, `pluck` six times faster than `map`.
```

Comment:
@mauro-oto I re-ran the test like this

```
2.1.2 :002 > puts Benchmark.measure { Order.where(number: ['R835174333', 'R551614735', 'R150666521', 'R333845264']).map &:id }
  Order Load (84.0ms)  SELECT `orders`.* FROM `orders` WHERE `orders`.`number` IN ('R835174333', 'R551614735', 'R150666521', 'R333845264') ORDER BY orders.created_at DESC
  0.080000   0.000000   0.080000 (  0.168243)
 => nil
2.1.2 :003 > exit
2.1.2 :002 > puts Benchmark.measure { Order.where(number: ['R835174333', 'R551614735', 'R150666521', 'R333845264']).pluck :id }
  SQL (1.1ms)  SELECT `orders`.`id` FROM `orders` WHERE `orders`.`number` IN ('R835174333', 'R551614735', 'R150666521', 'R333845264') ORDER BY orders.created_at DESC
  0.040000   0.000000   0.040000 (  0.047165)
 => nil
2.1.2 :003 > exit
```

And I got this results.   

```
  map:   0.080000   0.000000   0.080000 (  0.168243)
  pluck: 0.040000   0.000000   0.040000 (  0.047165)
```

This means that pluck is at least 3.5 times faster, not 6 times.